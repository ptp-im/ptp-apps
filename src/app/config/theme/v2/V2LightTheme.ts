import color from 'color';
import { MD2LightTheme } from 'react-native-paper';
import type { MD2Theme } from 'react-native-paper';

import { black, red800, white } from './colors';

export const V2LightTheme: MD2Theme = {
  ...MD2LightTheme,
  dark: false,
  roundness: 4,
  version: 2,
  isV3: false,
  colors: {
    ...MD2LightTheme.colors,
    // primary: '#6200ee',
    accent: '#03dac4',
    // background: '#f6f6f6',
    // background: grey50,
    surface: white,
    error: '#B00020',
    // text: black,
    // onSurface: '#000000',
    disabled: color(black).alpha(0.26).rgb().string(),
    placeholder: color(black).alpha(0.54).rgb().string(),
    backdrop: color(black).alpha(0.5).rgb().string(),
    // notification: pinkA400,

    msgUnreadBadge: red800,
    onSurface: 'rgb(30, 107, 192)',
    primary: 'rgb(30, 107, 192)',
    background: 'rgb(235, 235, 240)',
    white: 'rgb(255, 255, 255)',
    background1: 'rgb(235, 235, 240)',
    backgroundSelected: 'rgb(62, 125, 185)',
    backgroundTabs: 'rgb(255, 255, 255)',
    card: 'rgb(255, 255, 255)',
    text: 'rgb(28, 28, 30)',
    textPrimary: 'rgb(10, 132, 255)',
    textAlert: 'rgb(255, 59, 48)',
    border: 'rgb(229, 229, 229)',
    notification: 'rgb(255, 59, 48)',
  },
};

export const V2NavLightColors = {
  primary: V2LightTheme.colors.primary,
  background: V2LightTheme.colors.background,
  card: V2LightTheme.colors.card,
  text: V2LightTheme.colors.text,
  border: V2LightTheme.colors.border,
  notification: V2LightTheme.colors.notification,
};
