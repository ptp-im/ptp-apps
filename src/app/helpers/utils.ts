export function getUserIndexFromLocationUrl() {
  const location = window.location;
  const url = new URL(location.href);
  if (!url.searchParams.get('u') || !Number(url.searchParams.get('u'))) {
    return 0;
  } else {
    return Number(url.searchParams.get('u'));
  }
}

export function maskAddress(address: string) {
  return `${address.substring(0, 3)}**${address.substring(address.length - 3)}`;
}
