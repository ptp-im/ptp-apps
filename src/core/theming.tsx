import type { ColorValue } from 'react-native';

import { createTheming } from '@callstack/react-theme-provider';
import {
  argbFromHex,
  themeFromSourceColor,
} from '@material/material-color-utilities';
import color from 'color';

import {
  MD2DarkTheme,
  MD2LightTheme,
  MD3DarkTheme,
  MD3LightTheme,
} from '../styles/themes';
import { tokens } from '../styles/themes/v3/tokens';
import type {
  Theme,
  MD2Theme,
  MD3Theme,
  MD3Colors,
  MD3AndroidColors,
} from '../types';

export const DefaultTheme = MD3LightTheme;

const {
  ThemeProvider,
  withTheme,
  useTheme: useThemeProviderTheme,
} = createTheming<Theme>(DefaultTheme);

const useTheme = (overrides?: Parameters<typeof useThemeProviderTheme>[0]) =>
  useThemeProviderTheme<MD2Theme | MD3Theme>(overrides);

export { ThemeProvider, withTheme, useTheme };

export const defaultThemesByVersion = {
  2: {
    light: MD2LightTheme,
    dark: MD2DarkTheme,
  },
  3: {
    light: MD3LightTheme,
    dark: MD3DarkTheme,
  },
};

export const getTheme = (isDark = false, isV3 = true) => {
  const themeVersion = isV3 ? 3 : 2;
  const scheme = isDark ? 'dark' : 'light';

  return defaultThemesByVersion[themeVersion][scheme];
};

type Config = {
  sourceColor: ColorValue;
};

type Schemes = {
  lightScheme: MD3Colors;
  darkScheme: MD3Colors;
};

export const getDynamicThemeElevations = (scheme: MD3AndroidColors) => {
  const elevationValues = ['transparent', 0.05, 0.08, 0.11, 0.12, 0.14];
  return elevationValues.reduce((elevations, elevationValue, index) => {
    return {
      ...elevations,
      [`level${index}`]:
        index === 0
          ? elevationValue
          : color(scheme.surface)
              .mix(color(scheme.primary), elevationValue as number)
              .rgb()
              .string(),
    };
  }, {});
};

export const createDynamicThemeColors = ({ sourceColor }: Config): Schemes => {
  const { opacity } = tokens.md.ref;
  const modes = ['light', 'dark'] as const;

  const { schemes, palettes } = themeFromSourceColor(
    argbFromHex(color(sourceColor).hex())
  );

  const { light, dark } = modes.reduce(
    (prev, curr) => {
      const schemeModeJSON = schemes[curr].toJSON();

      const customColors = {
        surfaceDisabled: color(schemeModeJSON.onSurface)
          .alpha(opacity.level2)
          .rgb()
          .string(),
        onSurfaceDisabled: color(schemeModeJSON.onSurface)
          .alpha(opacity.level4)
          .rgb()
          .string(),
        backdrop: color(palettes.neutralVariant.tone(20))
          .alpha(0.4)
          .rgb()
          .string(),
      };

      const elevation = getDynamicThemeElevations(schemeModeJSON);

      const dynamicThemeColors = Object.assign(
        {},
        ...Object.entries(schemeModeJSON).map(([colorName, colorValue]) => ({
          [colorName]: color(colorValue).rgb().string(),
        })),
        {
          elevation,
          ...customColors,
        }
      );

      return {
        ...prev,
        [curr]: dynamicThemeColors,
      };
    },
    { light: MD3LightTheme.colors, dark: MD3DarkTheme.colors }
  );

  return { lightScheme: light, darkScheme: dark };
};
