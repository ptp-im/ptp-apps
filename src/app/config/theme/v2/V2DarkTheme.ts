import color from 'color';

import type { MD2Theme } from 'react-native-paper';
import { black, red800, white } from './colors';
import { V2LightTheme } from './V2LightTheme';

export const V2DarkTheme: MD2Theme = {
  ...V2LightTheme,
  dark: true,
  mode: 'adaptive',
  version: 2,
  isV3: false,
  colors: {
    ...V2LightTheme.colors,
    // primary: '#BB86FC',
    accent: '#03dac6',
    // background: '#121212',
    surface: '#121212',
    error: '#CF6679',
    onSurface: '#FFFFFF',
    // text: white,
    disabled: color(white).alpha(0.38).rgb().string(),
    placeholder: color(white).alpha(0.54).rgb().string(),
    backdrop: color(black).alpha(0.5).rgb().string(),
    // notification: pinkA100,

    msgUnreadBadge: red800,
    primary: 'rgb(10, 132, 255)',
    background: 'rgb(1, 1, 1)',
    card: 'rgb(18, 18, 18)',
    text: 'rgb(229, 229, 231)',
    textPrimary: 'rgb(10, 132, 255)',
    border: 'rgb(39, 39, 41)',
    notification: 'rgb(255, 69, 58)',
  },
};

export const V2NavDarkColors = {
  primary: V2LightTheme.colors.primary,
  background: V2LightTheme.colors.background,
  card: V2LightTheme.colors.card,
  text: V2LightTheme.colors.text,
  border: V2LightTheme.colors.border,
  notification: V2LightTheme.colors.notification,
};
