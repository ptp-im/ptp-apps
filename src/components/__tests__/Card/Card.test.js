import React from 'react';

import { render } from '@testing-library/react-native';
import color from 'color';
import renderer from 'react-test-renderer';

import { getTheme } from '../../../core/theming';
import { black, white } from '../../../styles/themes/v2/colors';
import Button from '../../Button/Button';
import Card from '../../Card/Card';
import { getCardColors } from '../../Card/utils';

describe('Card', () => {
  it('renders an outlined card', () => {
    const tree = renderer.create(<Card mode='outlined' />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders with a custom theme', () => {
    const { getByLabelText } = render(
      <Card
        mode='outlined'
        accessibilityLabel='card'
        theme={{ colors: { surface: '#0000FF' } }}
      />
    );

    expect(getByLabelText('card')).toHaveStyle({
      backgroundColor: '#0000FF',
    });
  });
});

describe('CardActions', () => {
  it('renders button with passed mode', () => {
    const { getByTestId } = render(
      <Card>
        <Card.Actions testID='card-actions'>
          <Button mode='contained'>Agree</Button>
        </Card.Actions>
      </Card>
    );

    expect(getByTestId('card-actions').props.children[0].props.mode).toBe(
      'contained'
    );
  });
});

describe('getCardColors - background color', () => {
  it('should return correct theme color, for theme version 3, contained mode', () => {
    expect(
      getCardColors({
        theme: getTheme(),
        mode: 'contained',
      })
    ).toMatchObject({ backgroundColor: getTheme().colors.surfaceVariant });
  });

  ['elevated', 'outlined'].forEach((mode) =>
    it(`should return correct theme color, for theme version 3, ${mode} mode`, () => {
      expect(
        getCardColors({
          theme: getTheme(),
          mode,
        })
      ).toMatchObject({ backgroundColor: getTheme().colors.surface });
    })
  );

  it('should return undefined, for theme version 2', () => {
    expect(
      getCardColors({
        theme: getTheme(false, false),
      })
    ).toMatchObject({ backgroundColor: undefined });
  });
});

describe('getCardColors - border color', () => {
  it('should return correct theme color, for theme version 3', () => {
    expect(
      getCardColors({
        theme: getTheme(),
      })
    ).toMatchObject({ borderColor: getTheme().colors.outline });
  });

  it('should return correct color, for theme version 2, dark mode', () => {
    expect(
      getCardColors({
        theme: getTheme(true, false),
      })
    ).toMatchObject({
      borderColor: color(white).alpha(0.12).rgb().string(),
    });
  });

  it('should return correct color, for theme version 2, light mode', () => {
    expect(
      getCardColors({
        theme: getTheme(false, false),
      })
    ).toMatchObject({
      borderColor: color(black).alpha(0.12).rgb().string(),
    });
  });
});
