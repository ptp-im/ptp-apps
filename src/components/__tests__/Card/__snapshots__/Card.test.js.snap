// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders an outlined card 1`] = `
<div
  className="css-view-175oi2r"
  dir={null}
  style={
    Object {
      "backgroundColor": "rgba(255,251,254,1.00)",
      "borderBottomLeftRadius": "4px",
      "borderBottomRightRadius": "4px",
      "borderTopLeftRadius": "4px",
      "borderTopRightRadius": "4px",
    }
  }
>
  <div
    className="css-view-175oi2r r-borderWidth-rs99b7 r-height-1pi2tsx r-position-u8s1d r-width-13qz1uu r-zIndex-136ojw6 r-pointerEvents-633pao"
    dir={null}
    style={
      Object {
        "borderBottomColor": "rgba(121,116,126,1.00)",
        "borderBottomLeftRadius": "4px",
        "borderBottomRightRadius": "4px",
        "borderLeftColor": "rgba(121,116,126,1.00)",
        "borderRightColor": "rgba(121,116,126,1.00)",
        "borderTopColor": "rgba(121,116,126,1.00)",
        "borderTopLeftRadius": "4px",
        "borderTopRightRadius": "4px",
      }
    }
  />
  <div
    aria-disabled={true}
    className="css-view-175oi2r r-flexGrow-16y2uox r-flexShrink-1wbh5a2"
    dir={null}
    onClick={[Function]}
    onContextMenu={[Function]}
    onKeyDown={[Function]}
    tabIndex="-1"
  />
</div>
`;
